crates/ide-assists/src/handlers/recursively_add_derive.rs

```diff
@@ -0,0 +1,1010 @@
+use crate::{AssistContext, Assists};
+use hir::{HasAttrs as _, HasCrate, HirFileId, ItemInNs};
+use ide_db::{
+    assists::{AssistId, AssistKind},
+    helpers::mod_path_to_ast,
+    imports::import_assets::NameToImport,
+    items_locator,
+};
+use smallvec::SmallVec;
+use std::fmt::Write;
+use syntax::{
+    ast::{self, edit_in_place::AttrsOwnerEdit, make, HasAttrs},
+    ted, AstNode, T,
+};
+
+// Assist: recursively_add_derive
+//
+// Recursively add `#[derive(...)]` attributes from a struct or enum, to the type of each field.
+// When the cursor is on a specific derive macro (such as `Copy`), only that derive will be added.
+// If the cursor is instead on the `derive` keyword itself, all derives that the struct or enum has will be added.
+//
+// ```
+// # //- minicore: derive, clone
+// struct X(u32);
+//
+// struct Y(u32);
+//
+// #[derive(Cl$0one)]
+// struct Point {
+//     x: X,
+//     y: Y,
+// }
+// ```
+// ->
+// ```
+// #[derive(Clone)]
+// struct X(u32);
+//
+// #[derive(Clone)]
+// struct Y(u32);
+//
+// #[derive(Clone)]
+// struct Point {
+//     x: X,
+//     y: Y,
+// }
+// ```
+pub(crate) fn recursively_add_derive(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    let attr = ctx.find_node_at_offset_with_descend::<ast::Attr>()?;
+    let path = attr.path()?;
+    let adt_src = ctx.find_node_at_offset::<ast::Adt>()?;
+    let adt = ctx.sema.to_def(&adt_src)?;
+    let adt_ty = adt.ty(ctx.db());
+
+    let current_module = ctx.sema.scope(adt_src.syntax())?.module();
+    let current_crate = current_module.krate();
+    let current_edition = current_crate.edition(ctx.db());
+
+    let derive_items = if path.syntax().text() == "derive" {
+        // The cursor is on the derive keyword, use all derive items the ADT has.
```

Comment:
Maybe add a comment explaining that we expand `#[derive(Foo, Bar)]` into `#[Foo] #[Bar]`, and so when we descend into macros if the cursor is at `Foo` the path won't be `derive` but `Foo` (it wasn't clear to me).

Reply:
Also a test for that is missing.

Reply:
That's a good idea, I do agree it's not very intuitive. I'm not entirely sure what to test for though, could you elaborate?

Reply:
Add a test that when there are multiple derives and the cursor is on one, only it is added.